Signal/src/views/PinEntryView.m

```diff
@@ -0,0 +1,205 @@
+    self.pinTextfield.secureTextEntry = YES;
+    self.pinTextfield.textAlignment = NSTextAlignmentCenter;
+    [self addSubview:self.pinTextfield];
+}
+
+- (UILabel *)createForgotLink
+{
+    UILabel *label = [UILabel new];
+    label.textColor = [UIColor ows_materialBlueColor];
+    NSString *text = NSLocalizedString(
+        @"REGISTER_2FA_FORGOT_PIN", @"Label for 'I forgot my PIN' link in the 2FA registration view.");
+    label.attributedText = [[NSAttributedString alloc]
+        initWithString:text
+            attributes:@{
+                NSForegroundColorAttributeName : [UIColor ows_materialBlueColor],
+                NSUnderlineStyleAttributeName : @(NSUnderlineStyleSingle | NSUnderlinePatternSolid)
+            }];
+    label.font = [UIFont ows_regularFontWithSize:ScaleFromIPhone5To7Plus(14.f, 16.f)];
+    label.numberOfLines = 0;
+    label.lineBreakMode = NSLineBreakByWordWrapping;
+    label.textAlignment = NSTextAlignmentCenter;
+    label.userInteractionEnabled = YES;
+    [label addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self
+                                                                        action:@selector(forgotPinLinkTapped:)]];
+    [self addSubview:label];
+    return label;
+}
+
+- (void)createSubmitButton
+{
+    const CGFloat kSubmitButtonHeight = 47.f;
+    // NOTE: We use ows_signalBrandBlueColor instead of ows_materialBlueColor
+    //       throughout the onboarding flow to be consistent with the headers.
+    OWSFlatButton *submitButton =
+        [OWSFlatButton buttonWithTitle:NSLocalizedString(@"REGISTER_2FA_SUBMIT_BUTTON",
+                                           @"Label for 'submit' button in the 2FA registration view.")
+                                  font:[OWSFlatButton fontForHeight:kSubmitButtonHeight]
+                            titleColor:[UIColor whiteColor]
+                       backgroundColor:[UIColor ows_signalBrandBlueColor]
+                                target:self
+                              selector:@selector(submitButtonWasPressed)];
+    self.submitButton = submitButton;
+    [self addSubview:submitButton];
+    [self.submitButton autoSetDimension:ALDimensionHeight toSize:kSubmitButtonHeight];
+}
+
+- (nullable NSString *)instructionsText
+{
+    return self.instructionsLabel.text;
+}
+
+- (void)setInstructionsText:(nullable NSString *)instructionsText
+{
+    self.instructionsLabel.text = instructionsText;
+}
+
+- (void)createContents
+{
+    const CGFloat kVSpacing = 30.f;
+
+    UILabel *instructionsLabel = [self createLabelWithText:nil];
+    self.instructionsLabel = instructionsLabel;
+    [instructionsLabel autoPinTopToSuperviewWithMargin:kVSpacing];
+    [instructionsLabel autoPinWidthToSuperview];
+
+    UILabel *createForgotLink = [self createForgotLink];
+    [createForgotLink autoPinEdge:ALEdgeTop toEdge:ALEdgeBottom ofView:instructionsLabel withOffset:5];
+    [createForgotLink autoPinWidthToSuperview];
```

Comment:
But we don't want a forgot link in the "set PIN settings" views, right?

Reply:
This isn't (currently?) used in the settings view. It's used in the registration and reminder views, both of which want a "I forgot my PIN" button.